### DISCUSSION TITLE
Can GitHub please provide an option to opt out from using a repository for AI training?

### DISCUSSION THREAD

**Discussion by gh-andre** [community/community#129511] [Copilot Conversations]
### Select Topic Area

Product Feedback

### Body

Some of us do not use Copilot and just want to host repositories, without them being used in the AI training process. Please provide an opt-out configuration option.

**Comment by islem-boudja** [comment]
I support this request for an opt-out option. As developers, we should have more control over how our code is used, especially when it comes to AI training.

A simple implementation could be a checkbox in repository settings or a special file (e.g., .github/NO_AI_TRAINING) that repository owners can use to indicate their preference. This would provide a clear, straightforward way for developers to express their wishes regarding AI training use of their code.

Implementing this option could increase trust and transparency in GitHub's practices, which is crucial for maintaining a positive relationship with the developer community. It respects the diverse opinions within the community about AI and machine learning in software development.

However, it's worth considering the potential impacts of such an option:

1. If widely adopted, it could affect the quality and capabilities of AI coding assistants like Copilot.
2. It might influence the open-source ecosystem, potentially creating a divide between AI-trainable and non-AI-trainable repositories.

Perhaps GitHub could also consider more granular controls, allowing repository owners to specify which parts of their code can be used for AI training, or provide clearer documentation on how hosted code is currently used in AI training processes.

Ultimately, giving developers more choice in how their contributions are used aligns with the spirit of open-source collaboration and could foster a more inclusive platform for all developers, regardless of their stance on AI in coding.

**Comment by TheFloatingBrain** [comment]
Strongly agree, there should be an opt out option, in addition to AI safety concerns, one might object to how a model is used or licensed.

**Comment by marcindulak** [comment]
For your information, there start appearing some unofficial (probably made up by AI) guides that claim GitHub offers such an opt-out. See https://github.com/redpanda-data/protoc-gen-go-mcp/pull/36#discussion_r2915033646

**Comment by gh-andre** [comment]
GitHub changed this setting a while ago. Now they took it further and it cannot be even changed from _Disabled_. I'll close this thread.

<img width="2733" height="1230" alt="Screenshot 2026-03-11 at 21-24-48 Copilot" src="https://github.com/user-attachments/assets/deff5b0f-8e07-4681-9258-6b154eb3ecf6" />

**Reply by marcindulak** [reply]
For reference, this is a global setting, available at https://github.com/settings/copilot/features "Settings -> Copilot -> Features"

**Reply by TheFloatingBrain** [reply]
I actually do not think this is what we asked for *exactly* this is for the use of Copilot when you are interacting with it *live*, I am not sure this applies to full repositories at rest (or retroactively)
From the  [FAQ](https://github.com/orgs/community/discussions/188488)
```
  What data are you collecting?

    When an individual user has this setting enabled, the interaction data we may collect includes:
        Outputs accepted or modified by the user
        Inputs sent to GitHub Copilot, including code snippets shown to the model
        Code context surrounding the user’s cursor position
        Comment and documentation that the user wrote
        File names, repository structure, and navigation patterns
        Interactions with Copilot features including Chat and inline suggestions
```

```

  I selected Copilot because GitHub said it didn’t train on user data. This feels like a bait and switch.

    As Copilot usage continues to increase dramatically, we’ve identified a need for real-world data to help our models cover the increasing number of scenarios they’re now being used for. We are committed to giving developers control over whether their interaction data is used for training and will always be transparent about our use of this data.
```


  I'm not sure this prevents Microsoft from using your repositories/contributions in full. But it does say
```
  You’re collecting code snippets, prompt text, AI responses, and detailed interaction patterns. How is this not giving you my entire codebase?

    We’re collecting data only from your interactions with Copilot—we are not pulling any data from your codebase at rest.

```
Reading through the FAQ that is not exactly clear, do they simply not have an option to prevent repo's from being used in full or does this imply the repo's are not ever (outside the live assistance at rest?

**Reply by gh-andre** [reply]
@TheFloatingBrain I think the FAQ point following the first one you mentioned is supposed to address all these forms of using our data with an explicit opt-out.

> * What can individuals do if they don’t want their inputs, outputs, or code snippets, used for model training?
> 
>   * They can opt out via their GitHub Copilot account settings at any time.

The option itself if poorly implemented and is not called _Allow_/_Disallow_, but instead _Enable_/_Disable_, which is ambiguous, but that's another issue.
